### ISSUE TITLE
Tor has successfully opened a circuit, but did not acknowledge serving haven's port 8888 despite being setup.

### ISSUE THREAD

**Issue by jidhub** [guardianproject/haven#290]
My tor browser[1] failed accessing Haven's logs[2] on the onion server hosted by "Orbot: proxy with tor" [3].
On the point of reporting this, I discovered[4] that Haven's servers was also accessible on localhost:8888. I suggested on a pull request[5] documenting that in the help text hovering in the app.

The server is running well on localhost:8.8.8.8, and it is now serving me content through a port forwarding.
I did never see orbot's warning "An app wants to open hidden server port 8888 to the Tor network. This is safe if you trust the app."[6].

[1]: linux64-7.5.3_en-US https://www.torproject.org/projects/torbrowser.html
[2]: 0.1.0-RC-1 installed with google play on android 6.0
[3]: 16.0.0-RC-2-multi-SDK16 also installed with google play on android 6.0; it says "NOTICE: Tor has successfully opened a circuit. Looks ..."
[4]: #230
[5] https://github.com/guardianproject/haven/pull/289
[6] https://github.com/guardianproject/haven/issues/175 ; it would be good to also document that message in a troubleshooting guide.

**Comment by lukeswitz** [comment]
In working on 2FA and multi device support, this part is without a doubt a headache. The proxy setup isn’t flexible...yet. 

The web server makes it all but easy to open a circuit, a big problem being the blank password with no feedback. It will ask Tor to open the port and then abort. Without enough time allotted, the call fails with no feedback and the user is confused. 

If you initially forgot a password I’d prefer it  alert the user instead of attempting to start the server. (The catch statement now is just logging there was blank input). 

I’m digging into the webserver code for a more stable solution so I can push these new features. An insecure server makes said features useless, thank you for the attention you’ve brought to this matter. Cheers,

Luke

**Comment by jidhub** [comment]
Tor browser did not show a password form, it just shows:

"Unable to connect

Firefox can’t establish a connection to the server at paXXXXXXXXus.onion:8888."

**Comment by lukeswitz** [comment]
Yes, I mean the password settings field in the Haven app itself.

**Comment by jidhub** [comment]
Sorry, can you explain why this password field, if functionning correctly when accessing localhost:8888, would no more work if using Tor ?

**Comment by lukeswitz** [comment]
Not definitively, but I can try before the bus gets to work:

- The server by design won’t start without a user supplied password. It will however attempt to start the server. (In that lapse, the timeout limit expires and the session is lost, but localhost may be accessible). 

- The calls to turn the server on and off sometimes require reopening the settings view after setting a password (this will be addressed also). 

- It serves an address unique to UDID and originates from your resolved localhost using port 8888. Problems with this configuration are known to exist in some countries with heavier restrictions on web traffic.

- When opening the port it’s assumed that the dialogue informed you of the local existence and therefore shouldn’t require extra documentation.

// TODO

- Address issues in protocol handling & inflexible responses to common configurations in Havens web server. 

- Save setting & adjust prefs on value changes.  

Without running a few tests I’m not much help; I’ll report back when I know more.

**Comment by n8fr8** [comment]
I think the issue here sounds like something at the Tor level, with @jidhub onion site ( paXXXXXXXXus.onion:8888) being unreachable.

**Comment by habib98** [comment]
wow
